// langchain/dist/vectorstores/mongodb_atlas.js
import { VectorStore } from "./base.js";
import { Document } from "../document.js";
import { maximalMarginalRelevance } from "../util/math.js";
export class MongoDBAtlasVectorSearch extends VectorStore {
    _vectorstoreType() {
        return "mongodb_atlas";
    }
    constructor(embeddings, args) {
        super(embeddings, args);
        Object.defineProperty(this, "collection", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        Object.defineProperty(this, "indexName", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        Object.defineProperty(this, "textKey", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        Object.defineProperty(this, "embeddingKey", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        this.collection = args.collection;
        this.indexName = args.indexName ?? "default";
        this.textKey = args.textKey ?? "text";
        this.embeddingKey = args.embeddingKey ?? "embedding";
    }
    async addVectors(vectors, documents) {
        const docs = vectors.map((embedding, idx) => ({
            [this.textKey]: documents[idx].pageContent,
            [this.embeddingKey]: embedding,
            ...documents[idx].metadata,
        }));
        await this.collection.insertMany(docs);
    }
    async addDocuments(documents) {
        const texts = documents.map(({ pageContent }) => pageContent);
        return this.addVectors(await this.embeddings.embedDocuments(texts), documents);
    }
    async similaritySearchVectorWithScore(query, k, filter) {
        const knnBeta = {
            vector: query,
            path: this.embeddingKey,
            k,
        };
        let preFilter;
        let postFilterPipeline;
        let includeEmbeddings;
        if (filter?.preFilter ||
            filter?.postFilterPipeline ||
            filter?.includeEmbeddings) {
            preFilter = filter.preFilter;
            postFilterPipeline = filter.postFilterPipeline;
            includeEmbeddings = filter.includeEmbeddings || false;
        }
        else
            preFilter = filter;
        if (preFilter) {
            knnBeta.filter = preFilter;
        }
        const pipeline = [
            {
                $search: {
                    index: this.indexName,
                    knnBeta,
                },
            },
            {
                $set: {
                    score: { $meta: "searchScore" },
                },
            },
        ];
        if (!includeEmbeddings) {
            const removeEmbeddingsStage = {
                $project: {
                    [this.embeddingKey]: 0,
                },
            };
            pipeline.push(removeEmbeddingsStage);
        }
        if (postFilterPipeline) {
            pipeline.push(...postFilterPipeline);
        }
        const results = this.collection.aggregate(pipeline);
        const ret = [];
        for await (const result of results) {
            const { score, [this.textKey]: text, ...metadata } = result;
            ret.push([new Document({ pageContent: text, metadata }), score]);
        }
        return ret;
    }
    /**
     * Return documents selected using the maximal marginal relevance.
     * Maximal marginal relevance optimizes for similarity to the query AND diversity
     * among selected documents.
     *
     * @param {string} query - Text to look up documents similar to.
     * @param {number} options.k - Number of documents to return.
     * @param {number} options.fetchK=20- Number of documents to fetch before passing to the MMR algorithm.
     * @param {number} options.lambda=0.5 - Number between 0 and 1 that determines the degree of diversity among the results,
     *                 where 0 corresponds to maximum diversity and 1 to minimum diversity.
     * @param {MongoDBAtlasFilter} options.filter - Optional Atlas Search operator to pre-filter on document fields
     *                                      or post-filter following the knnBeta search.
     *
     * @returns {Promise<Document[]>} - List of documents selected by maximal marginal relevance.
     */
    async maxMarginalRelevanceSearch(query, options) {
        const { k, fetchK = 20, lambda = 0.5, filter } = options;
        const queryEmbedding = await this.embeddings.embedQuery(query);
        // preserve the original value of includeEmbeddings
        const includeEmbeddingsFlag = options.filter?.includeEmbeddings || false;
        // update filter to include embeddings, as they will be used in MMR
        const includeEmbeddingsFilter = {
            ...filter,
            includeEmbeddings: true,
        };
        const resultDocs = await this.similaritySearchVectorWithScore(queryEmbedding, fetchK, includeEmbeddingsFilter);
        const embeddingList = resultDocs.map((doc) => doc[0].metadata[this.embeddingKey]);
        const mmrIndexes = maximalMarginalRelevance(queryEmbedding, embeddingList, lambda, k);
        return mmrIndexes.map((idx) => {
            const doc = resultDocs[idx][0];
            // remove embeddings if they were not requested originally
            if (!includeEmbeddingsFlag) {
                delete doc.metadata[this.embeddingKey];
            }
            return doc;
        });
    }
    static async fromTexts(texts, metadatas, embeddings, dbConfig) {
        const docs = [];
        for (let i = 0; i < texts.length; i += 1) {
            const metadata = Array.isArray(metadatas) ? metadatas[i] : metadatas;
            const newDoc = new Document({
                pageContent: texts[i],
                metadata,
            });
            docs.push(newDoc);
        }
        return MongoDBAtlasVectorSearch.fromDocuments(docs, embeddings, dbConfig);
    }
    static async fromDocuments(docs, embeddings, dbConfig) {
        const instance = new this(embeddings, dbConfig);
        await instance.addDocuments(docs);
        return instance;
    }
}
