// langchain/dist/vectorstores/xata.cjs
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.XataVectorSearch = void 0;
const base_js_1 = require("./base.cjs");
const document_js_1 = require("../document.cjs");
class XataVectorSearch extends base_js_1.VectorStore {
    _vectorstoreType() {
        return "xata";
    }
    constructor(embeddings, args) {
        super(embeddings, args);
        Object.defineProperty(this, "client", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        Object.defineProperty(this, "table", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        this.client = args.client;
        this.table = args.table;
    }
    async addDocuments(documents, options) {
        const texts = documents.map(({ pageContent }) => pageContent);
        return this.addVectors(await this.embeddings.embedDocuments(texts), documents, options);
    }
    async addVectors(vectors, documents, options) {
        const rows = vectors
            .map((embedding, idx) => ({
            content: documents[idx].pageContent,
            embedding,
            ...documents[idx].metadata,
        }))
            .map((row, idx) => {
            if (options?.ids) {
                return { id: options.ids[idx], ...row };
            }
            return row;
        });
        const res = await this.client.db[this.table].createOrReplace(rows);
        // Since we have an untyped BaseClient, it doesn't know the
        // actual return type of the overload.
        const results = res;
        const returnedIds = results.map((row) => row.id);
        return returnedIds;
    }
    async delete(params) {
        const { ids } = params;
        await this.client.db[this.table].delete(ids);
    }
    async similaritySearchVectorWithScore(query, k, filter) {
        const records = await this.client.db[this.table].vectorSearch("embedding", query, {
            size: k,
            filter,
        });
        // eslint-disable-next-line @typescript-eslint/no-explicit-any
        return records.map((record) => [
            new document_js_1.Document({
                pageContent: record.content,
                metadata: Object.fromEntries(Object.entries(record).filter(([key]) => key !== "content" &&
                    key !== "embedding" &&
                    key !== "xata" &&
                    key !== "id")),
            }),
            record.xata.score,
        ]);
    }
}
exports.XataVectorSearch = XataVectorSearch;
