// langchain/dist/document_loaders/web/azure_blob_storage_container.cjs
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.AzureBlobStorageContainerLoader = void 0;
const storage_blob_1 = require("@azure/storage-blob");
const azure_blob_storage_file_js_1 = require("./azure_blob_storage_file.cjs");
const base_js_1 = require("../base.cjs");
class AzureBlobStorageContainerLoader extends base_js_1.BaseDocumentLoader {
    constructor({ azureConfig, unstructuredConfig, }) {
        super();
        Object.defineProperty(this, "connectionString", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        Object.defineProperty(this, "container", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        Object.defineProperty(this, "unstructuredConfig", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        this.connectionString = azureConfig.connectionString;
        this.container = azureConfig.container;
        this.unstructuredConfig = unstructuredConfig;
    }
    async load() {
        const blobServiceClient = storage_blob_1.BlobServiceClient.fromConnectionString(this.connectionString);
        const containerClient = blobServiceClient.getContainerClient(this.container);
        let docs = [];
        for await (const blob of containerClient.listBlobsFlat()) {
            const loader = new azure_blob_storage_file_js_1.AzureBlobStorageFileLoader({
                azureConfig: {
                    connectionString: this.connectionString,
                    container: this.container,
                    blobName: blob.name,
                },
                unstructuredConfig: this.unstructuredConfig,
            });
            docs = docs.concat(await loader.load());
        }
        return docs;
    }
}
exports.AzureBlobStorageContainerLoader = AzureBlobStorageContainerLoader;
