// langchain/dist/retrievers/self_query/weaviate.cjs
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.WeaviateTranslator = void 0;
const ir_js_1 = require("../../chains/query_constructor/ir.cjs");
const base_js_1 = require("./base.cjs");
const utils_js_1 = require("./utils.cjs");
class WeaviateTranslator extends base_js_1.BaseTranslator {
    constructor() {
        super(...arguments);
        Object.defineProperty(this, "allowedOperators", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: [ir_js_1.Operators.and, ir_js_1.Operators.or]
        });
        Object.defineProperty(this, "allowedComparators", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: [
                ir_js_1.Comparators.eq,
                ir_js_1.Comparators.ne,
                ir_js_1.Comparators.lt,
                ir_js_1.Comparators.lte,
                ir_js_1.Comparators.gt,
                ir_js_1.Comparators.gte,
            ]
        });
    }
    formatFunction(func) {
        if (func in ir_js_1.Comparators) {
            if (this.allowedComparators.length > 0 &&
                this.allowedComparators.indexOf(func) === -1) {
                throw new Error(`Comparator ${func} not allowed. Allowed operators: ${this.allowedComparators.join(", ")}`);
            }
        }
        else if (func in ir_js_1.Operators) {
            if (this.allowedOperators.length > 0 &&
                this.allowedOperators.indexOf(func) === -1) {
                throw new Error(`Operator ${func} not allowed. Allowed operators: ${this.allowedOperators.join(", ")}`);
            }
        }
        else {
            throw new Error("Unknown comparator or operator");
        }
        const dict = {
            and: "And",
            or: "Or",
            eq: "Equal",
            ne: "NotEqual",
            lt: "LessThan",
            lte: "LessThanEqual",
            gt: "GreaterThan",
            gte: "GreaterThanEqual",
        };
        return dict[func];
    }
    visitOperation(operation) {
        const args = operation.args?.map((arg) => arg.accept(this));
        return {
            operator: this.formatFunction(operation.operator),
            operands: args,
        };
    }
    visitComparison(comparison) {
        if ((0, utils_js_1.isString)(comparison.value)) {
            return {
                path: [comparison.attribute],
                operator: this.formatFunction(comparison.comparator),
                valueText: comparison.value,
            };
        }
        if ((0, utils_js_1.isInt)(comparison.value)) {
            return {
                path: [comparison.attribute],
                operator: this.formatFunction(comparison.comparator),
                valueInt: parseInt(comparison.value, 10),
            };
        }
        if ((0, utils_js_1.isFloat)(comparison.value)) {
            return {
                path: [comparison.attribute],
                operator: this.formatFunction(comparison.comparator),
                valueNumber: parseFloat(comparison.value),
            };
        }
        throw new Error("Value type is not supported");
    }
    visitStructuredQuery(query) {
        let nextArg = {};
        if (query.filter) {
            nextArg = {
                filter: { where: query.filter.accept(this) },
            };
        }
        return nextArg;
    }
    mergeFilters(defaultFilter, generatedFilter, mergeType = "and") {
        if ((0, utils_js_1.isFilterEmpty)(defaultFilter?.where) &&
            (0, utils_js_1.isFilterEmpty)(generatedFilter?.where)) {
            return undefined;
        }
        if ((0, utils_js_1.isFilterEmpty)(defaultFilter?.where) || mergeType === "replace") {
            if ((0, utils_js_1.isFilterEmpty)(generatedFilter?.where)) {
                return undefined;
            }
            return generatedFilter;
        }
        if ((0, utils_js_1.isFilterEmpty)(generatedFilter?.where)) {
            if (mergeType === "and") {
                return undefined;
            }
            return defaultFilter;
        }
        const merged = {
            operator: "And",
            operands: [
                // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
                defaultFilter.where,
                // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
                generatedFilter.where,
            ],
        };
        if (mergeType === "and") {
            return {
                where: merged,
            };
        }
        else if (mergeType === "or") {
            merged.operator = "Or";
            return {
                where: merged,
            };
        }
        else {
            throw new Error("Unknown merge type");
        }
    }
}
exports.WeaviateTranslator = WeaviateTranslator;
